src/org/jetbrains/plugins/scala/codeInspection/collections/EmulateFlattenInspection.scala

```diff
@@ -0,0 +1,55 @@
+package org.jetbrains.plugins.scala.codeInspection.collections
+
+import org.jetbrains.plugins.scala.codeInspection.InspectionBundle
+import org.jetbrains.plugins.scala.extensions.ResolvesTo
+import org.jetbrains.plugins.scala.lang.psi.api.expr._
+import org.jetbrains.plugins.scala.lang.psi.api.statements.ScFunction
+
+/**
+  * @author Lukasz Piepiora
+  */
+class EmulateFlattenInspection extends OperationOnCollectionInspection {
+  override def possibleSimplificationTypes: Array[SimplificationType] =
+    Array(FlattenSimplification)
+}
+
+object FlattenSimplification extends SimplificationType {
+
+  override def hint: String = InspectionBundle.message("replace.with.flatten")
+
+  override def getSimplification(expr: ScExpression): Option[Simplification] =
+    expr match {
+      case seqOfSeqs `.flatMap` (identityOperation()) =>
+        Some(replace(expr).withText(invocationText(seqOfSeqs, "flatten")))
+      case _ => None
+    }
+
+  private object identityOperation {
+    def unapply(expr: ScExpression): Boolean = stripped(expr) match {
+      case identity(underscore()) => true
+      case identity() => true
+      case ScFunctionExpr(Seq(x), Some(ResolvesTo(param))) if x == param => true
+      case ScFunctionExpr(Seq(x), Some(identity(ResolvesTo(param)))) if x == param => true
+      case _ => false
+    }
+  }
+
+  private object identity {
+
```

Comment:
@niktrop Thanks for the tip! I've changed this, and added couple of tests for the new cases, which you pointed out wouldn't work in the original setup. However I have noticed something pretty strange.

Consider this test case:

```scala
  def testSuggests10(): Unit = {
    val selected = s"Set(Set(1), Set(2), Set(3)).${START}flatMap(identity _)$END"
    checkTextHasError(selected, hint, inspectionClass)
    val text = "Set(Set(1), Set(2), Set(3)).flatMap(identity _)"
    val result = "Set(Set(1), Set(2), Set(3)).flatten"
    testFix(text, result, hint)
  }
```

Which is pattern matched by this match (not exactly how I coded, but this is easier to see):

```scala
case seqOfSeqs `.flatMap` (underscore()) =>
```

Where I would expect it to be matched by something like this:

```scala
case seqOfSeqs `.flatMap` (`.identity`(_, underscore())) =>
```

Is my expectation incorrect, or there is some problem with the code. If it's the latter I would like to fix it, but I don't really know where, either PSI is incorrectly generated or the InvocationTemplate is parsing it wrong I guess. I think the PSI is incorrect but I'm not sure.

If I have 

```scala
flatMap(identity(_))
```

The PSI would be like this:

<img width="437" alt="screenshot 2016-04-04 21 22 00" src="https://cloud.githubusercontent.com/assets/105363/14260344/687bb20e-faab-11e5-9861-e06a495c464f.png">

Which is more or less logical especially in regards to `UnderscoreSection` which wraps `_`, and is placed in the argument list.

However when I have code like this:

```scala
flatMap(identity _)
```

The PSI will look like this:

<img width="410" alt="screenshot 2016-04-04 21 25 32" src="https://cloud.githubusercontent.com/assets/105363/14260483/fae6dd8a-faab-11e5-8b19-5167765ba82d.png">

Here the `UnderscoreSection spans from one parenthesis to another, and the identity call is wrapped in the `UnderscoreSection`. Not sure if this is correct, but if so, then this check:

```scala
object underscore {
    def unapply(expr: ScExpression): Boolean = {
      stripped(expr) match {
        case ScParenthesisedExpr(underscore()) => true
        case typed: ScTypedStmt if typed.expr.isInstanceOf[ScUnderscoreSection] => true
        case und: ScUnderscoreSection => true
        case _ => false
      }
    }
  }
```

Seems to be too relaxed in the case `case und: ScUnderscoreSection => true`

Is my thinking correct?